packages/wdio-browserstack-service/src/insights-handler.ts

```diff
@@ -0,0 +1,471 @@
+        }
+
+        const req = this.requestQueueHandler.add(uploadData)
+
+        if (req.proceed) {
+            await uploadEventData(req.data, req.url)
+        }
+    }
+
+    private async sendTestRunEventForCucumber (world: ITestCaseHookParameter, eventType: string) {
+        const uniqueId = getUniqueIdentifierForCucumber(world)
+
+        let testMetaData = this._tests[uniqueId]
+        if (!testMetaData) testMetaData = {}
+
+        const { feature, scenario, steps } = testMetaData
+
+        let fullNameWithExamples = world.pickle.name
+        const examples = getScenarioExamples(world)
+        if (examples) {
+            fullNameWithExamples = world.pickle.name + ' (' + examples.join(', ')  + ')'
+        }
+
+        /* istanbul ignore next */
+        let testData: TestData = {
+            uuid: testMetaData.uuid,
+            started_at: testMetaData.startedAt,
+            finished_at: testMetaData.finishedAt,
+            type: 'test',
+            body: {
+                lang: 'webdriverio',
+                code: null
+            },
+            name: fullNameWithExamples,
+            scope: fullNameWithExamples,
+            scopes: [feature?.name || ''],
+            identifier: scenario?.name,
+            file_name: feature?.path,
+            vc_filepath: (this._gitConfigPath && feature?.path) ? path.relative(this._gitConfigPath, feature?.path) : undefined,
+            location: feature?.path,
+            framework: this._framework,
+            result: 'pending',
+            meta: {
+                feature: feature,
+                scenario: scenario,
+                steps: steps,
+                examples: examples
+            }
+        }
+
+        if (eventType == 'TestRunStarted') {
+            testData['integrations'] = {}
+            if (this._browser && this._platformMeta) testData['integrations'][getCloudProvider(this._browser)] = this.getIntegrationsObject()
+        }
+
+        /* istanbul ignore if */
+        if (world.result) {
+            let result: string = world.result.status.toLowerCase()
+            if (result !== 'passed' && result !== 'failed') result = 'skipped' // mark UNKNOWN/UNDEFINED/AMBIGUOUS/PENDING as skipped
+            testData.finished_at = (new Date()).toISOString()
+            testData.result = result
+            testData.duration_in_ms = world.result.duration.nanos / 1000000 // send duration in ms
+
+            if (result == 'failed') {
+                testData.failure = [
+                    {
+                        'backtrace': [world.result.message ? removeAnsiColors(world.result.message) : 'unknown']
+                    }
+                ],
+                testData.failure_reason = world.result.message ? removeAnsiColors(world.result.message) : world.result.message,
+                testData.failure_type = world.result.message == undefined ? null : world.result.message.toString().match(/AssertionError/) ? 'AssertionError' : 'UnhandledError'
```

Comment:
is calling `toString()` really necessary? Do we expect `message` to be something else than a string?
```suggestion
                if (world.result.message) {
                    testData.failure_type = world.result.message.toString().match(/AssertionError/)
                        ? 'AssertionError'
                        : 'UnhandledError'
                }
```

Reply:
> is calling toString() really necessary?

correct, not needed. removing